### ISSUE TITLE
Not an issue: Does verlok lazy load support Async loading?

### ISSUE THREAD

**Issue by Niresh12495** [verlok/lazyload#57]
If it supports then it would be great

**Comment by verlok** [comment]
@Niresh12495 what do you exactly mean with Async loading?

**Comment by Niresh12495** [comment]
In all your examples, lazy load js library is loaded before the options but will it work
in the following case because with async loading, images can render later when ever the lazyload  js has been loaded (like in the case load images after everything has been loaded)
In the example below im using different data_src and data_srcset because you can't check if async loading working with default values
`
<img alt="Sneakers &amp; Tennis shoes basse" 
					data-lazy="../img/44738717AM_15_a.jpg" 
					data-lazyset="../img/44738717AM_15_a.jpg 1x, ../img/44738717AM_11_a.jpg 2x">
<script>var myLazyLoad = new LazyLoad({
    data_src: "lazysrc",
    data_srcset: "lazyset",
});
</script>
<script src=/path/to/lazyload.js"></script>`
so will lazyload work in this case you have to check yourself

**Comment by Niresh12495** [comment]
When you execute something synchronously, you wait for it to finish before moving on to another task. When you execute something asynchronously, you can move on to another task before it finishes.

If lazy load is async it can be loaded anytime or loaded with condition

**Comment by Niresh12495** [comment]
Any reply? or updates?

**Comment by verlok** [comment]
Sorry @Niresh12495 I've been very busy with my actual job and beyond that preparing the talk for CSS Day 2017. I look forward to be back starting next Monday.

**Comment by verlok** [comment]
I still fail to understand how would it work with the code you posted above. If you instantiate `new LazyLoad()` before declaring the `LazyLoad` function, wouldn't it generate an error in any case?

**Comment by Niresh12495** [comment]
`<script async src="//pagead2.googlesyndication.com/
pagead/js/adsbygoogle.js"></script>

<ins class="adsbygoogle"
    style="display:inline-block;width:728px;height:90px"
    data-ad-client="ca-pub-xxxxxxxxxxxxxxxx"
    data-ad-slot="1234567890"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>`

Just look at this adsense code, when ever the adsbygoogle.js gets loaded, it executes `(adsbygoogle = window.adsbygoogle || []).push({});`  and then ads gets loaded
that is how the async works, adsbygoogle.js can be loaded any time (after page render) or any where  (at footer)

**Comment by verlok** [comment]
Ok what you wrote in your latest message is not 100% correct, but I see what you mean. 
Your final goal would be to set your LazyLoad options anywhere (just like `adsbygoogle`) and then load LazyLoad asynchronously, without having to wait for it to initialize it.
That's definetly a good idea. 

What about this?

```html
<!-- images markup -->
<img alt="Sneakers &amp; Tennis shoes basse" 
					data-lazysrc="../img/44738717AM_15_a.jpg" 
					data-lazyset="../img/44738717AM_15_a.jpg 1x, ../img/44738717AM_11_a.jpg 2x">

<!-- LazyLoad configuration -->
<script>
window.lazyLoadOptions = {
    data_src: "lazysrc",
    data_srcset: "lazyset"
}
</script>
<!-- LazyLoad async download and auto-execution -->
<script src="/path/to/lazyload.js" async defer></script>
```

It currently wouldn't work like that, but it wouldn't be a great deal for me to update it for it to work like that.

What do you think?

**Comment by verlok** [comment]
Hey @Niresh12495, I've worked on it right now, and I've released the fix to make it work like I wrote above.
Check out the edit on the `feature/asyncInit` branch. I've added a demo under the `demos` folder, named `async.html`, and a note on the `README.md` file.

Let me know if it works for you.

**Comment by Niresh12495** [comment]
You understood what i meant 😄  Cool
but i have tried your new script not seems to be working
[Testing Async](http://codepen.io/Niresh12495/pen/yMqzVE)
Correct me if I'm wrong
Thanks...

**Comment by Niresh12495** [comment]
[Code Pen Test Here](http://codepen.io/Niresh12495/pen/yMqzVE)
When i remove the async attribute it stops working

**Comment by verlok** [comment]
@Niresh12495 

That's why you put the options after the script. The options need to be before the script, in case it loads very fast (from the browser cache, for example).

Here's what you should have in the Async Lazy Test.html

```html
<head>
	<script>
	window.asyncLazyLoadOptions = { 
	    throttle: 150,
	    skip_invisible: false,
	    threshold: 150,
		data_src: 'lazysrc',
	    data_srcset: 'lazyset'
	};
	</script>
	<script async src="https://rawgit.com/verlok/lazyload/feature/asyncInit/src/lazyload.js"></script>
</head>

<body>
	<img data-lazysrc="https://raw.githubusercontent.com/verlok/lazyload/feature/asyncInit/img/44704882BQ_11_a.jpg" />
</body>
```

I just tested it, and it worked.

**Comment by verlok** [comment]
[Codepen test here](http://codepen.io/verlok/pen/WpKMoX)

**Comment by verlok** [comment]
Now I'm thinking about adding the ability to have more than one asynchronously loaded LazyLoad.
Maybe with something like:

```javascript
window.asyncLazyLoadOptions = [{
    throttle: 150,
    skip_invisible: false,
    threshold: 150,
    data_src: 'lazysrc',
    data_srcset: 'lazyset'
}, {
    container: '#aScrollingDiv',
    threshold: 150, 
    data_src: 'lazysrc',
    data_srcset: 'lazyset'
}];
```

I think it would be a great idea to preserve the ability to have multiple LazyLoad on page, yet load the script asynchronously.

**Comment by Niresh12495** [comment]
`That's why you put the options after the script. The options need to be before the script, in case it loads very fast (from the browser cache, for example).`

In async the script can be loaded anyhow?

Here is an example just make sure to pause your adblock before clicking 
[AsyncTest.html.zip](https://github.com/verlok/lazyload/files/870645/AsyncTest.html.zip)

So will your script work if it is called onclick?

**Comment by verlok** [comment]
I think you don't understand what async means exactly, Niresh.

The order of the `script` tags is important even if a script is loaded asynchronously, because `async` means that you cannot know when the script is executed, which could also be "right away" if the script is dowloaded and executed very fast (i.g. from cache), that's why the configuration variable needs to have a value before requesting the script.

To answer your question: yes, my script would work on click in the case you just wrote reported. Here's the [AsyncTest.html loading LazyLoad upon click](http://codepen.io/verlok/pen/NpBYVK?editors=1000).

**Comment by Niresh12495** [comment]
> If you aren't dynamically loading scripts or marking them as defer or async, then scripts are loaded in the order encountered in the page. It doesn't matter whether it's an external script or an inline script - they are executed in the order they are encountered in the page. Inline scripts that come after external scripts have are held until all external scripts that came before them have loaded and run.
> 
> Async scripts (regardless of how they are specified as async) load and run in an unpredictable order. The browser loads them in parallel and it is free to run them in whatever order it wants.

This is an answer from 
[Stack overflow](http://stackoverflow.com/questions/8996852/load-and-execute-order-of-scripts)

If your script works with async and doesnt work without async then it is a bug
Please read the link
Thanks

**Comment by Niresh12495** [comment]
Ok i have made a small change in your lazyload.js : now it works with or without the async tag

[lazyload.js.zip](https://github.com/verlok/lazyload/files/870719/lazyload.js.zip)

> Now I'm thinking about adding the ability to have more than one asynchronously loaded LazyLoad.

This is something which i really need because
i use the following data_src values: for <img tags i use data-lazysrc and  for background images i use data-background

Thanks 😄

**Comment by verlok** [comment]
I don't want to force the users to attach lazyload on the `window.load` event, so I won't add your change, also because the version of my script under the `feature/async` branch already works with or without `async`.

I will continue to implement this feature allowing users to pass an array of options.
Thank you for your inputs.

**Comment by Niresh12495** [comment]
So why it doesn't work here
[Bug](http://codepen.io/Niresh12495/pen/XMBqzV)

**Comment by verlok** [comment]
Jeez, you're impossible :)
It doesn't work there because you didn't put the `async` attribute in the script.

```html
<!-- WRONG async usage -->
<script src="https://rawgit.com/verlok/lazyload/feature/asyncInit/src/lazyload.js"></script>
<!-- CORRECT async usage -->
<script src="https://rawgit.com/verlok/lazyload/feature/asyncInit/src/lazyload.js" async></script>
```

**Comment by Niresh12495** [comment]
Google analytics works with or without async attribute,
Google Adsense works with or without async attribute,
Twitter's widget.js works with or without  async attribute
i still wonder why your lazyload doesn't work with async attribute?

> 
> <script src="https://rawgit.com/verlok/lazyload/feature/asyncInit/src/lazyload.js"></script>

I know this is wrong async usage, but an async script should work with or without async attribute, something is wrong think about it.

**Comment by verlok** [comment]
It would work without `async` either, if you just put the global options object before the script, as I told you some posts ago. :)

**Comment by verlok** [comment]
In addition, to make LazyLoad listen to the `window`'s `load` event would be a very bad practice for my script, because it's users might want control on its exact execution time, maybe because they need to integrate it with others scripts. 

In fact, the examples you reported (Google analytics etc) are script that don't have any influence on the web page / web app execution, they just do additional things of their own.

**Comment by verlok** [comment]
Released version 6.2.0 which now supports async loading and auto initialitazion.

**Comment by verlok** [comment]
@Niresh12495 please read the `async` loading note on my README.md, including the note about DOM readiness. Thanks.
